db-engine/src/main/java/com/splicemachine/db/impl/sql/compile/BinaryComparisonOperatorNode.java

```diff
@@ -150,28 +150,38 @@ public ValueNode bindExpression(FromList fromList,
 		if (! leftTypeId.isStringTypeId() && rightTypeId.isStringTypeId())
 		{
 			DataTypeDescriptor rightTypeServices = rightOperand.getTypeServices();
-
-			rightOperand =  (ValueNode)
-				getNodeFactory().getNode(
-					C_NodeTypes.CAST_NODE,
-					rightOperand, 
-					new DataTypeDescriptor(
-							rightTypeId,
-							true, 
-							rightTypeServices.getMaximumWidth()),
-					getContextManager());
+			TypeId targetTypeId;
+			if (leftTypeId.isBooleanTypeId() || leftTypeId.isDateTimeTimeStampTypeId())
+				targetTypeId = leftTypeId;
+			else
+				targetTypeId = rightTypeId;
+
+			rightOperand = (ValueNode)
+					getNodeFactory().getNode(
+							C_NodeTypes.CAST_NODE,
+							rightOperand,
+							new DataTypeDescriptor(
+									targetTypeId,
+									true,
+									rightTypeServices.getMaximumWidth()),
```

Comment:
For the case where we need to cast to the left type, should we use the left type's typeServices.getMaximumWidth()?